File: core/src/com/alta_v2/physics/task/TaskCreationManager.java
package com.alta_v2.physics.task;

import com.alta_v2.physics.executionContext.TiledMapEngineContext;
import com.alta_v2.physics.executionContext.reserveData.ReservableActor;
import com.alta_v2.physics.task.moveNpc.MoveNpcChecker;
import com.alta_v2.physics.task.moveNpc.MoveNpcTask;
import com.alta_v2.physics.task.movePlayer.MovePlayerChecker;
import com.alta_v2.physics.task.movePlayer.MovePlayerTask;
import com.alta_v2.physics.task.rotatePlayer.RotateNpcChecker;
import com.alta_v2.physics.task.rotatePlayer.RotatePlayerChecker;
import com.alta_v2.physics.task.rotatePlayer.RotatePersonTask;
import com.alta_v2.physics.utils.MovementCalculator;
import com.alta_v2.rendering.tiledMapScreen.actors.PersonView;
import com.badlogic.gdx.math.Vector2;
import lombok.experimental.UtilityClass;
import lombok.extern.log4j.Log4j2;

/**
 * Provides mediator to tasks creation manage.
 */
@Log4j2
@UtilityClass
public final class TaskCreationManager {

    /**
     * Creates the task that calculates movement of player.
     *
     * @param direction - the direction of movement.
     * @param context   - the engine execution context.
     * @return crated {@link ResultTiledMapTask} instance or {@code}null{@code} if failed creation.
     */
    public ResultTiledMapTask createPlayerMoveTask(MovementDirection direction, TiledMapEngineContext context) {
        MovePlayerChecker checker = new MovePlayerChecker();

        if (!checker.canTaskBeExecuted(context)) {
            log.trace("Failed to perform movement of focus point on the {} because process is blocked", direction);
            return null;
        }

        Vector2 targetPoint = MovementCalculator.getTargetPointLocal(
                direction, context.getFocusPointLocal().getX(), context.getFocusPointLocal().getY()
        );
        if (!MovementCalculator.canMoveTo(targetPoint.x, targetPoint.y, context.getAltitudeMap())) {
            log.debug("Failed to run player moving since [{}: {}] is blocked", targetPoint.x, targetPoint.y);
            return null;
        }

        return MovePlayerTask.builder()
                .altitudeMap(context.getAltitudeMap())
                .direction(direction)
                .focusPointGlobal(context.getFocusPointGlobal())
                .focusPointLocal(context.getFocusPointLocal())
                .playerPointLocal(context.getPlayer().getLocalPoint())
                .playerView(context.getPlayer().getView())
                .isPlayerMoving(context.getPlayer().getIsMoving())
                .targetPointLocal(targetPoint)
                .build();
    }

    /**
     * Creates the task to rotate player.
     *
     * @param direction - the direction of movement.
     * @param context   - the engine execution context.
     * @return crated {@link ResultTiledMapTask} instance or {@code}null{@code} if failed creation.
     */
    public ResultTiledMapTask createRotatePlayerTask(MovementDirection direction, TiledMapEngineContext context) {
        PersonView currentView = MovementDirection.getPersonView(direction);
        return new RotatePlayerChecker(currentView).canTaskBeExecuted(context)?
                new RotatePersonTask(direction, context.getPlayer().getView()) : null;
    }

    /**
     * Creates the task that calculates movement of NPC.
     *
     * @param direction - the direction of movement.
     * @param npcId     - the ID of NPC to be moved.
     * @param context   - the engine execution context.
     * @return created {@link ResultTiledMapTask} instance or {@code}null{@code} if creation failed.
     */
    public ResultTiledMapTask createNpcMoveTask(MovementDirection direction, int npcId, TiledMapEngineContext context) {
        MoveNpcChecker checker = new MoveNpcChecker(npcId);
        if (!checker.canTaskBeExecuted(context)) {
            log.trace("Failed to perform movement of Npc {} to the {} because process is blocked", npcId, direction);
            return null;
        }

        ReservableActor npc = context.getNpcMap().get(npcId);
        Vector2 targetPoint = MovementCalculator.getTargetPointLocal(
                direction, npc.getLocalPoint().getX(), npc.getLocalPoint().getY()
        );
        if (!MovementCalculator.canMoveTo(targetPoint.x, targetPoint.y, context.getAltitudeMap())) {
            log.debug("Failed to run npc moving since [{}: {}] is blocked", targetPoint.x, targetPoint.y);
            return null;
        }

        return MoveNpcTask.builder()
                .direction(direction)
                .npc(npc)
                .targetPointLocal(targetPoint)
                .altitudeMap(context.getAltitudeMap())
                .focusPointGlobal(context.getFocusPointGlobal())
                .build();
    }

    /**
     * Creates the task to rotate player.
     *
     * @param direction - the direction of movement.
     * @param npcId     - the ID of NPC to be moved.
     * @param context   - the engine execution context.
     * @return crated {@link ResultTiledMapTask} instance or {@code}null{@code} if failed creation.
     */
    public ResultTiledMapTask createRotateNpcTask(int npcId, MovementDirection direction, TiledMapEngineContext context) {
        PersonView currentView = MovementDirection.getPersonView(direction);
        return new RotateNpcChecker(npcId, currentView).canTaskBeExecuted(context) ?
                new RotatePersonTask(direction, context.getNpcMap().get(npcId).getView()) : null;
    }

}

File: core/src/com/alta_v2/physics/task/TaskCreationManager.kt
package com.alta_v2.physics.task

import com.alta_v2.physics.executionContext.TiledMapEngineContext
import com.alta_v2.physics.task.moveNpc.MoveNpcChecker
import com.alta_v2.physics.task.moveNpc.MoveNpcTask
import com.alta_v2.physics.task.movePlayer.MovePlayerChecker
import com.alta_v2.physics.task.movePlayer.MovePlayerTask
import com.alta_v2.physics.task.rotatePlayer.RotateNpcChecker
import com.alta_v2.physics.task.rotatePlayer.RotatePersonTask
import com.alta_v2.physics.task.rotatePlayer.RotatePlayerChecker
import com.alta_v2.physics.utils.MovementCalculator
import mu.KotlinLogging

/**
 * Provides mediator to tasks creation manage.
 */
class TaskCreationManager {

    companion object {

        private val log = KotlinLogging.logger {  }

        /**
         * Creates the task that calculates movement of player.
         *
         * @param direction - the direction of movement.
         * @param context   - the engine execution context.
         * @return crated [ResultTiledMapTask] instance or ``null`` if failed creation.
         */
        fun createPlayerMoveTask(direction: MovementDirection, context: TiledMapEngineContext): ResultTiledMapTask? {
            val checker = MovePlayerChecker()
            if (!checker.canTaskBeExecuted(context)) {
                log.trace("Failed to perform movement of focus point on the {} because process is blocked", direction)
                return null
            }
            val targetPoint = MovementCalculator.getTargetPointLocal(
                    direction, context.focusPointLocal.x, context.focusPointLocal.y
            )
            if (!MovementCalculator.canMoveTo(targetPoint.x, targetPoint.y, context.altitudeMap)) {
                log.debug("Failed to run player moving since [{}: {}] is blocked", targetPoint.x, targetPoint.y)
                return null
            }

            return MovePlayerTask(
                    playerId = context.player.id,
                    playerPointLocal = context.player.localPoint,
                    playerView = context.player.view,
                    isPlayerMoving = context.player.isMoving,
                    altitudeMap = context.altitudeMap,
                    direction = direction,
                    focusPointGlobal = context.focusPointGlobal,
                    focusPointLocal = context.focusPointLocal,
                    targetPointLocal = targetPoint
            )
        }

        /**
         * Creates the task to rotate player.
         *
         * @param direction - the direction of movement.
         * @param context   - the engine execution context.
         * @return crated [ResultTiledMapTask] instance or ``null`` if failed creation.
         */
        fun createRotatePlayerTask(direction: MovementDirection, context: TiledMapEngineContext): ResultTiledMapTask? {
            val currentView = MovementDirection.getPersonView(direction)
            return if (RotatePlayerChecker(currentView).canTaskBeExecuted(context)) RotatePersonTask(direction, context.player.view) else null
        }

        /**
         * Creates the task that calculates movement of NPC.
         *
         * @param direction - the direction of movement.
         * @param npcId     - the ID of NPC to be moved.
         * @param context   - the engine execution context.
         * @return created [ResultTiledMapTask] instance or ``null`` if creation failed.
         */
        fun createNpcMoveTask(direction: MovementDirection, npcId: Int, context: TiledMapEngineContext): ResultTiledMapTask? {
            val checker = MoveNpcChecker(npcId)
            if (!checker.canTaskBeExecuted(context)) {
                log.trace("Failed to perform movement of Npc {} to the {} because process is blocked", npcId, direction)
                return null
            }
            val npc = context.npcMap[npcId]
            val targetPoint = MovementCalculator.getTargetPointLocal(
                    direction, npc!!.localPoint.x, npc.localPoint.y
            )
            if (!MovementCalculator.canMoveTo(targetPoint.x, targetPoint.y, context.altitudeMap)) {
                log.debug("Failed to run npc moving since [{}: {}] is blocked", targetPoint.x, targetPoint.y)
                return null
            }

            return MoveNpcTask(
                    direction = direction,
                    npc = npc,
                    targetPointLocal = targetPoint,
                    altitudeMap = context.altitudeMap,
                    focusPointGlobal = context.focusPointGlobal
            )
        }

        /**
         * Creates the task to rotate player.
         *
         * @param direction - the direction of movement.
         * @param npcId     - the ID of NPC to be moved.
         * @param context   - the engine execution context.
         * @return crated [ResultTiledMapTask] instance or ``null`` if failed creation.
         */
        fun createRotateNpcTask(npcId: Int, direction: MovementDirection, context: TiledMapEngineContext): ResultTiledMapTask? {
            val currentView = MovementDirection.getPersonView(direction)
            return if (RotateNpcChecker(npcId, currentView).canTaskBeExecuted(context)) RotatePersonTask(direction, context.npcMap[npcId]!!.view) else null
        }
    }
}